compiler/rustc_abi/src/layout.rs

```diff
@@ -307,12 +337,11 @@ impl<Cx: HasDataLayout> LayoutCalculator<Cx> {
                         common_non_zst_abi_and_align = Err(AbiMismatch);
                     } else {
                         // Fields with the same non-Aggregate ABI should also
-                        // have the same alignment
-                        if !matches!(common_abi, Abi::Aggregate { .. }) {
-                            assert_eq!(
-                                common_align, field.align.abi,
-                                "non-Aggregate field with matching ABI but differing alignment"
-                            );
```

Comment:
This assert is similar to the one in the sanity check: Hitting it means that an invalid layout was already created elsewhere.

Instead of trying to make the consumers of layouts work with invalid layouts, I think we should find the place where the invalid layouts are created to begin with and emit the `LayoutCalculatorError::ReprConflict` there.

Reply:
Philosophical discussion about the meaning of ICEs aside, that does result in a much smaller fix. I feel a bit silly for heading down that path as far as I did. Thanks!